DOC 2126: Writing a CorDapp using a template tutorials rewrite (#483)
corda/corda-docs: content/en/docs/corda-os/4.8/run-your-cordapp.md

        }
        rpcUsers = [[ user: "user1", "password": "test", "permissions": ["ALL"]]]
    }
    node {
        name "O=PartyB,L=New York,C=US"
        p2pPort 10008
        rpcSettings {
            address("localhost:10009")
            adminAddress("localhost:10049")
        }
        rpcUsers = [[ user: "user1", "password": "test", "permissions": ["ALL"]]]
    }
}
```

2. Click the right-hand side **Gradle** tab and open the **cordformation** folder. Double-click **deployNodes**.

   For each node definition, Gradle will:

   * Package the project’s source files into a CorDapp `.jar`.
   * Create a new node in `build/nodes` with your CorDapp already installed.

## Run the nodes

Running `deployNodes` builds the nodes under `build/nodes`. If you navigate to one of these folders, you'll see
the three node folders. Each node folder has the following structure:

```bash
.
|____additional-node-infos
|____certificates
|____corda.jar                  // The runnable node.
|____cordapps
|____djvm
|____drivers
|____logs
|____network-parameters
|____node.conf                // The node's configuration file.
|____nodeInfo
|____persistence.mv.db
|____persistence.trace.db

```

> 1. In terminal, navigate to `build/nodes` from the root directory.

Author: Specified that this is the IntelliJ terminal window.

2. Start the nodes by running the following command:

   * Mac OSX: `runnodes`
   * Windows: `runnodes.bat`

This will start a terminal window for each node. Give each node a moment to start - you’ll know it’s ready when its terminal windows displays
the message “Welcome to the Corda interactive shell.”.

{{< figure alt="running node" zoom="/en/images/running_node.png" >}}

## Start the IOU flow

> Now that our nodes are running, order one of them to create an IOU by kicking off our `IOUFlow`. In a larger

Author: Already updated with Ben's changes.

app, you’d generally provide a web API sitting on top of our node. Here, for simplicity, you’ll be interacting with the
node via its built-in CRaSH shell.

1. Go to the terminal window displaying the CRaSH shell of `PartyA`.

2. Type `help` to display a list of the available commands.

   {{< note >}}
   Local terminal shell is available only in a development mode. In a production environment SSH server can be enabled.
   More about SSH and how to connect can be found on the [Node shell](shell.md) page.
   {{< /note >}}

> 3. You want to create an IOU of 99 with `PartyB`. To start the `IOUFlow`, type the following syntax:

Author: Updated

   ```bash
   start IOUFlow iouValue: 99, otherParty: "O=PartyB,L=New York,C=US"
   ```

   This single command will cause `PartyA` and `PartyB` to automatically agree an IOU. This is one of the great advantages of

>    the flow framework - it allows you to reduce complex negotiation and update processes into a single function call.

Author: No idea but I don't think it matters for the purposes of this tutorial.

        borrower: "O=PartyB, L=New York, C=US"
        contract: "com.template.contracts.TemplateContract"
        notary: "O=Notary, L=London, C=GB"
        encumbrance: null
        constraint: !<net.corda.core.contracts.SignatureAttachmentConstraint>
         key: "aSq9DsNNvGhYxYyqA9wd2eduEAZ5AXWgJTbTEw3G5d2maAq8vtLE4kZHgCs5jcB1N31cx1hpsLeqG2ngSysVHqcXhbNts6SkRWDaV7xNcr6MtcbufGUchxredBb6"
     ref:
      txhash: "D189448F05D39C32AAAAE7A40A35F4C96529680A41542576D136AEE0D6A80926"
      index: 0
    statesMetadata:
    - ref:
       txhash: "D189448F05D39C32AAAAE7A40A35F4C96529680A41542576D136AEE0D6A80926"
       index: 0
      contractStateClassName: "com.template.states.IOUState"
      recordedTime: "2020-10-19T11:09:58.183Z"
      consumedTime: null
      status: "UNCONSUMED"
      notary: "O=Notary, L=London, C=GB"
      lockId: null
      lockUpdateTime: null
      relevancyStatus: "RELEVANT"
      constraintInfo:
        constraint:
          key: "aSq9DsNNvGhYxYyqA9wd2eduEAZ5AXWgJTbTEw3G5d2maAq8vtLE4kZHgCs5jcB1N31cx1hpsLeqG2ngSysVHqcXhbNts6SkRWDaV7xNcr6MtcbufGUchxredBb6"
      totalStatesAvailable: -1
      stateTypes: "UNCONSUMED"
      otherResults: []
      ```

>       This is the transaction issuing our `IOUState` onto a ledger.

Reviewer: ```suggestion
      This is the transaction issuing your `IOUState` onto a ledger.
```

5. If you run the same command on the other node (the notary), you will see the following:

   ```bash
   {
     "states" : [ ],
     "statesMetadata" : [ ],
     "totalStatesAvailable" : -1,
     "stateTypes" : "UNCONSUMED",
     "otherResults" : [ ]
   }
   ```

   This is the result of Corda’s privacy model. Because the notary was not involved in the transaction and had no need to see the data, the
   transaction was not distributed to them.

## Conclusion

You have written a simple CorDapp that allows IOUs to be issued onto the ledger. This CorDapp is made up of two key
parts:

> * The `IOUState`, which represents IOUs on the blockchain.

Author: Should be "on the ledger." Updated.

* The `IOUFlow`, which orchestrates the process of agreeing the creation of an IOU on-ledger.

After completing this tutorial, your CorDapp should look like this:

* Java: [https://github.com/corda/corda-tut1-solution-java](https://github.com/corda/corda-tut1-solution-java)
* Kotlin: [https://github.com/corda/corda-tut1-solution-kotlin](https://github.com/corda/corda-tut1-solution-kotlin)

## Next steps

> There are a number of improvements you could make to this CorDapp:
> 
> * You could add unit tests, using the contract-test and flow-test frameworks.
> * You could change `IOUState.value` from an integer to a proper amount of a given currency.
> * You could add an API, to make it easier to interact with the CorDapp.

Reviewer 2: I'd suggest moving the 'you could' to the lead-in sentence for efficiency:

```suggestion
You could make several improvements to this CorDapp. You could:

* Add unit tests, using the contract-test and flow-test frameworks.
* Change `IOUState.value` from an integer to a proper amount of a given currency.
* Add an API, to make it easier to interact with the CorDapp.
```

But for now, the biggest priority is to add an `IOUContract` imposing constraints on the evolution of each

> `IOUState` over time - see [Applying contract constraints](tut-two-party-introduction.md).

Author: Updated
